### ISSUE TITLE
Allow colorbar.ax.set_ylim to set the colorbar limits?

### ISSUE THREAD

**Issue by ImportanceOfBeingErnest** [matplotlib/matplotlib#13948]
### Bug report

**Bug summary**

Coming from [this stackoverflow question](https://stackoverflow.com/questions/55665167/asymmetric-color-bar-with-fair-diverging-color-map) I was wondering if it would be useful to allow the limits of the colorbar axes to be used for restricting the range of colors to be shown in the colorbar. I.e.

    colorbar.ax.set_ylim(-2, 8) 

**Code for reproduction**

```
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors

x = np.linspace(-2,8)

mima = np.max((np.abs(x.min()), np.abs(x.max())))
norm = mcolors.Normalize(-mima, +mima)

sc = plt.scatter(x,x, c=x, cmap = 'seismic', norm=norm)

cb = plt.colorbar(sc, orientation="horizontal")
cb.ax.set_ylim(x.min(), x.max())

plt.margins(x=0)
plt.savefig("test.png", bbox_inches="tight")
plt.show()
```

**Actual outcome**

![test](https://user-images.githubusercontent.com/23121882/56083247-a87a5800-5e22-11e9-8c87-ba4510396c70.png)

**Expected outcome**

![test](https://user-images.githubusercontent.com/23121882/56083268-d790c980-5e22-11e9-9419-90919ae00ab2.png)

**Comment by jklymak** [comment]
What’s supposed to happen if limits are set smaller? Also what version are you using above?

**Comment by timhoffm** [comment]
On master, your code produces
 
![grafik](https://user-images.githubusercontent.com/2836374/56084083-124c2f00-5e2e-11e9-95d0-2cf8d5e0f0fe.png)

It seems, you have used `cb.ax.set_xlim`.

I don't know what the suggested way of working with colorbar is right now. However, manipulating `xlim/ylim` of the underlying axes seems too low level. Probably. Colorbar should have a dedicated method for the effect you want to achieve.

**Comment by jklymak** [comment]
You can't do this just by changing the x-limit of the underlying axes because the colorbar is represented by a `QuadMesh`, and the whole quadmesh will still get shown regardless of the underlying xlim.  

Its totally possible to add this extra API, but I'm skeptical whether its worth the substantial complexity, versus just truncating the colormap.

**Comment by ImportanceOfBeingErnest** [comment]
I ran this with master (fetched ~2 weeks ago, `3.0.2.post2051+g628ab4e23`). 
I think when limits are smaller, it should just behave as usual axes, limiting what's shown in the axes. I suppose the more interesting question is, what should happen if limits are larger? That could be up for debate, but I would be open for it to either show the complete colormap (due to the norm going from -8 to 8) or to show white (due to the mappable array being limited between -2 and 8).

**Comment by jklymak** [comment]
If anyone takes this up please remember that the colorbars also have the extend arrows on the end.  This would have to interact with those properly. 

However I’ll again say that I’m not sure the API complexity needed is worth it.

**Comment by efiring** [comment]
I share @jklymak's skepticism about cost/benefit ratio for this.  Colorbars are already pretty complicated.

**Comment by spalato** [comment]
Supporting this change.
Condiser the following motivating example:
```import numpy as np
import matplotlib.pyplot as plt

x = np.linspace(-10, 10)
y = np.linspace(-10, 10)

z = np.exp(-0.5*(x[:,np.newaxis]**2/10.0+y[np.newaxis,:]**2/2.0)) 
noise = 0.1*np.random.randn(*z.shape)
z += noise

lim = np.max(np.abs(z))

plt.pcolormesh(x, y, z, vmin=-lim, vmax=lim, cmap="RdBu_r")
plt.colorbar(ax=None)
```
![index](https://user-images.githubusercontent.com/7659022/69246354-dee23780-0ba8-11ea-992b-529c9f1a3c04.png)

In this image, a diverging colormap is used to differentiate positive and negative values. The negative values are due to some noise.

The colorbar falsely conveys the information that the lower limit of the data is comparable to its upper limit. It also requires lots of unnecessary space, which is often a problem in real contexts.

Using `set_clim` shifts the white color to the center of the new range, defeating the purpose of the diverging colormap.

> I share @jklymak's skepticism about cost/benefit ratio for this. Colorbars are already pretty complicated.

I couldn't find any workaround. Currently, my solution is to avoid displaying the colorbar altogether, or to post-process in a vector graphics software.

**Comment by ImportanceOfBeingErnest** [comment]
I think people aknowledge the problem, but there is no definitive solution to it. Possible solutions are:

1. Change the limits of the colormap (this issue)
2. Allow for a normalization that cuts the colormap (#15333)
3. Provide a colorbar factory (as discussed [on discourse](https://discourse.matplotlib.org/t/limiting-colormapping/20598/4))

The [discourse thread](https://discourse.matplotlib.org/t/limiting-colormapping/20598/4)  also contains workarounds using solutions 2. and 3. Those should allow users to get the desired functionality.  
The remaining question here is whether a solution should be incorporated into matplotlib, and if so, which one.

**Comment by jklymak** [comment]
I strongly disagree that the colorbar falsely conveys that there is just as much negative as positive data in the above plot.  That’s the image’s job. The colorbar shown above shows me quite clearly that a properly centred diverging colormap has been used, and indicates the range that has been assigned to each side of zero. the data shows me that most of the data is greater than zero.  If the ranges were assymetric I’d need to search around for zero to make sure it was properly zeroed (you may or may not be surprised by how many people don’t bother to centre diverging colormaps) and then I’d need to squint to make sure the colorscale for the two signs is symmetric.

**Comment by AdityaSavara** [comment]
While I actually would prefer this feature, my understanding is that somebody could convey what was desired by making a custom color map. I believe that's what is discussed here: https://stackoverflow.com/questions/41916204/specify-boundaries-on-matplotlib-colorbar

**Comment by jklymak** [comment]
This is mostly closed by #20054, but there should be tests if we decide to keep this as a "feature"...
